Major cleanup

nene opened the pull request:
After several pull requests to clean up the large `Editor.tsx` file, I concluded that it's going to take me ages if I try to do it through small easily reviewable pull requests. So here's a humongous one instead 😱 

It's still a work-in-progress, but here are the major changes so far:

- Workout chart components are now all named Bars: `SteadyBar`, `RampBar`, `FreeBar`, `RepetitionBar`.
- Instead of using single `Bar` interface with lots of optional fields, there are now four separate interfaces with mandatory fields: `SteadyInterval`, `RampInterval`, `FreeInterval`, `RepetitionInterval`.
- For now I completely dropped support for distance based workouts. All this distance v/s time calculation logic was very cumbersome and hard to un-entangle. I concluded that it's simpler to just delete it and re-implement from scratch. So now the size of each interval-block is always determined by its duration, which simplifies things a lot.
- Converted all plain .js files to TypeScript
- Multiple new top-level dirs:
    - `src/storage/` abstracts away localStorage access
    - `src/xml/` contains ZWO reading and writing
    - `src/types/` is a home for various widely used types.
    - `src/interval/` has helpers for creating `Interval` objects and manipulating intervals array.
    - `src/network/` contains upload and download helpers.
- Extracted bunch of new components: `PaceSelector`, `Head`, `Keyboard`, `Notification`, `Stats`, `Title`, `NumberField` & multiple `Button` components.
- Lots of renaming. Notably:
    - `.time` is now `.duration`.
    - `actionId` is now `selectedId`
- Deleted `package-lock.json` as there is a `yarn.lock` anyway (and one shouldn't really use multiple lock files).
- Added `.editorconfig`.
- Pulled in some missing `@types/*` libraries.

As a result of it all, `Editor.tsx` is now half the size it was previously.

One thing I have been struggling with is the Firebase integration. I'm not at all familiar with Firebase and don't even know where to look for to learn about this (Firebase seems to be many things). Also I have no idea if and how I could test locally that the Firebase-related code still works.

breiko83 commented:
Hey @nene what do you mean with "I completely dropped support for distance based workouts"? 
This is one of the key features of this editor and for running workouts on Zwift. How do you intend to re-implement it?

nene commented:
Basically I see two options:

- Always storing the duration value, but converting it to distance for display and ZWO output when in distance-mode.
- Storing either duration or distance depending on the current mode, but never both at the same time. (When mode gets switched, converting all durations to distances or vice versa).

The first option is simpler to implement, while the second option might allow setting the distances more precisely. (Like one could want to set distance to 100 meters. Given speed of 5.8 m/s, this will result in duration = 17.2413 seconds. That in turn would result in distance of 99.9995 meters. I the error will be small enough to not really matter, but not 100% sure.)

I think I'll go with the first option and see if it has any drawbacks.

nene commented:
Truth be told, I've never used Zwift for running, and as so, I know little about the run workouts. So if you can give me some guidance on the differences from cycling workouts, it's much appreciated.

breiko83 commented:
Same as converting bike to run workouts it doesn’t make sense converting
distance to time workouts for running.

I honestly would not bother implementing a conversion.

In terms of distance based workouts Zwift only allows a minimum block
distance (and step) of 200 meters. This is why you can find the round to
200 in the code.

nene commented:
Thanks for the tips.

Didn't do much today... was pretty wasted after Zwift Fondo.

Implemented switching between bike/run with the cleanup of existing workout. Also did some improvements on `<Label>`, so it now also uses the Interval objects and run workouts no more use FTP value to calculate intensity percentage, which felt quite weird.

nene commented:
Another batch of changes...

- Renamed `power` to `intensity` - the power field didn't actually contain the wattage, as the name "power" might indicate.
- Renamed `<Comment/>` to `<InstructionEditor/>` - to better align with the name of `Instruction` interface.
- Added concept of a Mode, which replaces passing around of 3 parameters (`ftp`, `weight`, `sportType`) with just a single `mode` parameter.  For start there are two modes:
    - BikeMode contains `ftp` & `weight` - provides methods for calculating power & wkg.
    - RunMode contains `runningTimes` - provides methods for calculating distance.
- Reorganized the `runningTimes` from object of time-strings to array of plain seconds. Makes it easier to work with the data. Not sure it's the final form.
- Bunch of fiddling with `calculateEstimatedTimes()`. For now kept the original algorithm (see #37), but simplified it.

The distances are now shown again for run workouts (with the caveat of #38).

I now need to get the distance-based workouts done.

nene commented:
A few additional things:

- Added Ramda library for some useful utility function
- Moved functions in `helpers.ts` to more specific files: `src/util/duration.ts` and `src/util/math.ts`
- Added helpers for distance calculation: `src/util/distance.ts`
- Showing the total distance for run workouts
- Hiding of TSS for run workouts (fixes #39)

nene commented:
After some fiddling around concluded that my first approach doesn't really work out as I thought it would. I really need to keep the distance as base for distance-based workouts, not compute it from duration.

So I'll go and try if I can manage to toggle between time and duration as the base measure.

In the mean time, some other changes:

- Brought back DurationType toggle (which doesn't do much at the moment)
- Moved duration-to-width and intensity-to-height calculations to base Mode class - so this logic is now kept in a single place, instead of scattered around in three Bar components.
- Extracted multiple sub-components from RampBar to make it simpler.
- Removed the sizing-from-middle functionality of ramps. To me this seemed like unnecessary functionality, but YMMV, I can easily add it back later. These height1, height2, height3 things confused the hell out of me. I had to get rid of one to be able to dissect it.

nene commented:
Finally the distance-based workouts now work again.

- There's now a `Length` type, which is either `Distance` or `Duration`.
- Intervals contain `.length` fields which store either of these depending on current editor setting.
- Renamed `DistanceType` to `LengthType` to be in line with this new naming.
- Switching between time & distance modes clears the existing workout.
- Most of the magic happens inside Mode class, where `widthToLength()` and `lengthToWidth()` helpers perform conversion between pixel-size and Duration/Distance (depending on the selected lengthType).

There are a few bugs which I've spotted here and there (and likely some I haven't spotted), but in general the data-flow inside the app is more consistent. In the old implementation I've spotted many scenarios where the change in one place doesn't get properly reflected in another part (e.g. you a change the 1 mile time field, but neither distance or time changes in the Label-popup of an interval).

One area which I know isn't backwards compatible, is reading the previous workout from localStorage. If you try to load a workout saved with earlier version of the app, things will likely break. Not sure if it's worth to create a conversion script or simply skip loading of the old workout when incompatible.

Another area which is completely untested is the whole Firebase integration. As I noted before, I know fairly little of Firebase.

nene commented:
Fixed various bugs in XML parsing and generation. Plus added tests.

This part of the code really needs a proper cleanup. Especially the parsing can be simplified a whole lot.

nene commented:
Some general thoughts:

- Should really adopt a state management like [Redux](https://redux.js.org/). Currently several options (like current sportType and lengthType) need to be passed deep inside various components, which complicates things quite a bit. Keeping all this state inside Editor also adds to the responsibilities of this component, making the file larger. The operations of modifying the intervals add/delete/move/etc would be a great fit for a reducer.
- I have steered away from touching CSS in this big update. So the names of components and CSS classes have somewhat diverged and the CSS and related components aren't necessarily always close together. My personal choice would be [styled-components](https://styled-components.com/) lib, though there are also other options to keep a tighter connection between CSS and React components.
- XML processing is too complex. For some reason two different libraries are used for reading and writing XML, although `xml-js` can accomplish both.

nene commented:
In the mean time I've added bunch of additional tests:

- the `Label` component is now mostly covered,
- `RepetitionBar` also finally has some tests.
- Did some mocking to avoid re-testing SteadyBar components inside it, and mocked out `uuid` lib to avoid variations in ID-s in tests.
- Covered `Mode` classes and `duration-`, `distance-`, `interval-` utils.

With tests safety-net in place, refactored a few more things:

- Moved `intervalDistance()` and `intervalDuration()` helpers inside Mode classes.
- Extracted formatting-related functionality to separate `format` util. The methods in Mode for calculating speed, distance etc now always compute things using all the available precision, any rounding will happen in `format` util.
- Made use of helpers from Ramda to transform arrays without mutation with less code.

Also a few tiny UI tweaks:

- `Kg` -> `kg` (which is the proper SI unit).
- `100w, 20W/kg, 200 m` -> `100 w, 20 W/kg, 200 m` there's now always a space between number and its unit.

nene commented:
Hey, there. Any hope from you reviewing this?

nene commented:
I looked a bit at this Firebase integration code and realized that currently loading the old workouts definitely doesn't work because they are saved as simple snapshots of the program data, and as I changed the whole representation of intervals, trying to load the old stuff is bound to blow up.

But I see that the workouts are also saved as ZWO files. So instead of writing conversion from old workout data structure to new, I think it's better to just read the ZWO files instead. The workout metadata (name, description, length, etc) can still be stored in Firebase like before - so it could be easily accessed in workouts list. Also better to avoid duplication, and the ZWO format is likely more future-proof than the current JSON.

Thoughts?

nene commented:
Two reasons:

- The workout is saved as a ZWO file to AWS anyway, so why save it twice?
- I have considerably changed the in-app workout data format, so loading the existing workouts (saved in old format) would not work without writing some conversion layer. But that didn't seem like a worthwhile thing to implement given that conversion to & from ZWO already exists.

breiko83 commented:
I actually need the workout data to load the workouts saved on Firebase. There is no reading/parsing from S3.

nene commented:
Emm... I'm not sure what you mean.

- There definitely is code in `master` branch to fetch and parse ZWO data from S3: https://github.com/breiko83/zwo-editor/blob/c1723adc9e5255e1e4a61e73195eb5505ed033ae/src/components/Editor/Editor.tsx#L655
- In my version, I've reused this code to replace the reading of workout from Firebase.

Or do I misunderstand your question?

nene commented:
FYI: As I've lost hope of getting these changes into this repo, I will be doing a fork.

breiko83 commented:
Rene -

Your PR is currently breaking core functionalities and it can’t be merged.

If you want to contribute in a way that doesn’t change software but
improves it I’m happy to merge. Otherwise feel free to have your own fork.


On Mon, 1 Feb 2021 at 18:34, Rene Saarsoo <notifications@github.com> wrote:

> FYI: As I've lost hope of getting these changes into this repo, I will be
> doing a fork.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/breiko83/zwo-editor/pull/36#issuecomment-771066708>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAESBV7K2A2PBGB2LPQPFLLS43X3BANCNFSM4VRGYDBA>
> .
>

nene commented:
But could you explain which core functionalities it's breaking?

I suppose it's related to not saving full workout data to Firebase, which, as I said above, seemed redundant and unnecessary to me. I might be mistaken, but as I can't test this part of the app by myself (or at least haven't figured out how to do so), I can only guesstimate whether the my changes break the Firebase / AWS / etc integrations or not. You hinted that something is broken regarding this, but I've been unable to understand what exactly.

What other functionalities are getting broken?

breiko83 commented:
Saving data on Firebase was also for (future) visualisation purposes.
Reading from file is not as fast as reading from db.

Also workouts parser for running isn’t developed. So if data is not saved
on Firebase it can’t be restored from S3.

As I said this is core functionality and I can’t merge something that
breaks it.

On Mon, 1 Feb 2021 at 19:15, Rene Saarsoo <notifications@github.com> wrote:

> But could you explain which core functionalities it's breaking?
>
> I suppose it's related to not saving full workout data to Firebase, which,
> as I said above, seemed redundant and unnecessary to me. I might be
> mistaken, but as I can't test this part of the app by myself (or at least
> haven't figured out how to do so), I can only guesstimate whether the my
> changes break the Firebase / AWS / etc integrations or not. You hinted that
> something is broken regarding this, but I've been unable to understand what
> exactly.
>
> What other functionalities are getting broken?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/breiko83/zwo-editor/pull/36#issuecomment-771092481>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAESBVY4M2AY54ZLEX55MM3S434WJANCNFSM4VRGYDBA>
> .
>

nene commented:
> workouts parser for running isn’t developed.

Well, I made the workouts parser also work for running, so this should no more be the problem.

> Saving data on Firebase was also for (future) visualisation purposes
> Reading from file is not as fast as reading from db.

So, you're saying that currently it doesn't actually break any functionality? Right?

Anyway... what do you say if I:

- Restore the save-workout-to-firebase logic - writing it in the new format.
- Add a new "version" field to the workouts in firebase, to be able to easily distinguish between new and old format.
- Implement a way to read the existing workouts from Firebase, which were written there in old format.

The downside of this would be that we would end up with a mixture of workouts in Firebase: some saved in new- while others in old format. This same problem would pop up again whenever the internal workout data-structure happens to change. Having these multiple formats would in turn complicate any future processing of this workout data in Firebase. One could overcome this by running a conversion script to convert all workouts in Firebase to new format, after which the backwards-compatibility code could be dropped from the app.

I personally see lots of downsides with this approach. If you think it's fine though, I'd implement it. Or perhaps you have some alternative solution in mind?

breiko83 commented:
Sorry Rene - I need more thought on where this is heading.

I'm a bit confused wether this open source editor should have S3 and Firebase implemented at all. This is not you fault - just my initial decision that I guess it's not appropriate for an open source project.

Maybe I should fork my own version (for zwiftworkout.com) and we should remove the login/registration/saving functionalities from this one. What do you think?

nene commented:
Well, forking is likely not the best path forward, as forks are hard to maintain.

What you might want instead is to separate the pure editor component into a library. That would though still leave bunch of other parts one would need to interact with this editor:

- Converting to/from ZWO
- Adding of different types of intervals
- Handling the rest of the workout data (name, author, description, ...)
- Switches for toggling sportType and lengthType
- Editors for FTP, weight, running times.
- Calculations of stats like total duration and TSS.

So... separating out purely the editor is complicated, to say the least.

A bit simpler option would be to separate out pretty much the whole editor page, leaving out the server-related buttons. Though then it'll be less flexible with regards to how one can add these additional buttons. However, at least it sounds more feasible way than the previous option.

nene commented:
FYI, I went ahead with my original forking plan.

My fork is now published here: https://nene.github.io/zwo-editor/

For now it doesn't really provide any extra functionality besides bugfixes, and in fact lots of functionality has been completely dropped (e.g. the whole server-side stuff). On the surface the app is still pretty similar, though internally, there's probably only a few lines I haven't touched.

breiko83 commented:
No problem! Thanks for your support and good luck with everything.
